src/qibolab/unrolling.py

```diff
@@ -0,0 +1,79 @@
+"""Utilities for sequence unrolling.
+
+May be reused by different instruments.
+"""
+
+from dataclasses import asdict, dataclass, field, fields
+from functools import total_ordering
+
+from .pulses import PulseSequence
+
+
+def _waveform(sequence: PulseSequence):
+    # TODO: deduplicate pulses (Not yet as drivers may not support it yet)
+    # TODO: count Rectangular and delays separately (Zurich Instruments supports this)
+    # TODO: Any constant part of a pulse should be counted only once (Zurich Instruments supports this)
+    # TODO: check if readout duration is faithful for the readout pulse (I would only check the control pulses)
+    # TODO: Handle multiple qubits or do all devices have the same memory for each channel ?
+    return sequence.duration - sequence.ro_pulses.duration
+
+
+def _readout(sequence: PulseSequence):
+    # TODO: Do we count 1 readout per pulse or 1 readout per multiplexed readout ?
+    return len(sequence.ro_pulses)
+
+
+def _instructions(sequence: PulseSequence):
+    return len(sequence)
+
+
+@total_ordering
+@dataclass(frozen=True, eq=True)
+class Bounds:
+    """Instument memory limitations proxies."""
+
+    waveforms: int = field(metadata={"count": _waveform})
+    """Waveforms estimated size."""
+    readout: int = field(metadata={"count": _readout})
+    """Number of readouts."""
```

Comment:
> I would suggest having a bit more elaborate explanation here. E.g. in case of ZI instruments, there is a limitation on how many results (complex numbers in general) can be stored in the device memory. The number of things that get stored in that memory is related to the number of readouts but is not necessarily equal.

I definitely agree:
https://github.com/qiboteam/qibolab/pull/817#discussion_r1505600303
but at the moment that's what it is.

In any case, we could comment what it should be, instead of what it is (as for the waveforms memory above). We could leave the discussion about "what it is" to the computing method, that can declare both what is supposed to be, and then enter the details.
The only problem is that the computing method is private, so I wonder if instead we should comment everything in the attributes... (that to me seems misplaced, but it has the advantage of being exposed)
https://github.com/qiboteam/qibolab/blob/78b5103af3eb20fd9d7f29485176d55a19857837/src/qibolab/unrolling.py#L21-L23

Reply:
If I remenber correctly @GabrielePalazzo we found in the past some limitation before that was around 30 readout pulses with 4096 shots. Way below the result limitation, I think it was due to uploading the readout pulse several times and the memory getting filled ? Anyhow, we could try to keep both bounds or just change this one to one related to the results memory.